File: ProjectEuler/CPP/p10.cc
//author agis daniels
//goal is Find the sum of all the primes below two million.?
#include <iostream> // cin cout
#include <math.h> // sqrt
#include <bitset> // for bit set
#include <vector> // for the vector

using namespace std;

//is_prime(int n)
//this is a function that can be used uses trivial division
//param n is the number to check
//return true if prime false otherwise
bool is_prime(int n){
    if(!(n%3)){
        return false;
    }
    int r = sqrt(n);
    for(int f = 5 ; f <= r;f+=6){
        if(n%f==0 || n%(f+2)==0){
            return false;
        }
    }
    return true;
}

//get_sieve of eratosthenes(const int size)
//sieve of eratosthenes the bool part of it
//param size is the size we would like to cover
//return the seive in a bit set
bitset<1000000> get_sieve_of_eratosthenes(const int size){
    bitset<1000000> seive;
    const int size2 = size/2;
    const int r     = (sqrt(size)-1)/2;
    seive.set(0);
    for(int i=1; i<=r; ++i){
        if(0==seive[i]){
            for(int j=i*(i+1)+i*(i+1); j<size2; j+=(2*i+1)){
                seive.set(j);
            }
        }
    }
    return seive;
}

//get_sieve(const int size)
//sieve of eratosthenes the vector part
//param size used for the size of the sieve
//return the seive in a vector
vector<int> get_seive(const int size){
    bitset<1000000> seive = get_sieve_of_eratosthenes(size);
    vector<int> rVec;
    rVec.push_back(2);
    for(int i=999999; i>0; --i){
        if(0==seive[i]){
            rVec.push_back(i+i+1);
        }
    }
    return rVec;
}

int main(){
    vector<int> vec_seive = get_seive(2000000);
    long sum_primes    = 0;
    int prime_set_size = vec_seive.size();
    for(int i = prime_set_size-1 ; i >= 0 ; --i){
        sum_primes += vec_seive[i];
    }
    cout << "sum of first 2000000 is " << sum_primes << '.' <<  endl;
    return 0;
}

File: ProjectEuler/Python/p10.py
from math import sqrt

def is_prime(n):
  if n == 1:   return False
  if n < 4:    return True
  if 0 == n%2: return False
  if n < 9:    return True
  if 0 == n%3: return False
  lim = int(sqrt(n)+2)
  for i in range(5, lim, 6):
    if 0 == n%i or 0 == n%(i+2):
      return False
  return True

def sol1(lim):
  ans = sum([i for i in range(3, lim, 2) if is_prime(i)])+2
  print('sum of all the primes below {} is {}'.format(lim, ans))

def gen_bit_sieve(lim):
  lim2 = lim // 2
  r = int(sqrt(lim)-1)//2
  bit_sieve = [False]*lim
  for i in range(1, r+1):
    if not bit_sieve[i]:
      for j in range(i*(i+1)+i*(i+1), lim2, 2*i+1):
        bit_sieve[j] = True
  return bit_sieve

def gen_sieve(lim):
  bit_sieve = gen_bit_sieve(lim)
  sieve = [2] + [i+i+1 for i in range((lim//2)-1, 0, -1) if not bit_sieve[i]]
  return sieve

def sol2(lim):
  ''' a refined solution for the trivial version '''
  print('sum of all the primes below {} is {}'.format(lim, sum(gen_sieve(lim))))

def main():
  sol2(2000000)
  sol1(2000000)

main()
 
